prov/usnic/src/usdf_fabric.c

```diff
@@ -623,6 +595,96 @@ usdf_get_distance(
     return ret;
 }
 
+/* Check all things related to a device. Make sure it's okay, the source address
+ * matches the requested address, the destination is reachable from the device,
+ * the device fabric name matches the requested fabric name, and the device
+ * domain name matches the requested domain name.
+ *
+ * @param version Libfabric API version used to verify the domain / fabric name.
+ * @param hints   Hints passed to fi_getinfo.
+ * @param src     Source address being requested.
+ * @param dest    Destination address to communicate with.
+ * @param dep     usNIC device entry being checked.
+ *
+ * @return true on success, false on failure. For debug logging can be enabled
+ *         to see why a device was disqualified.
+ */
+static bool usdf_check_device(uint32_t version, struct fi_info *hints,
+			      struct sockaddr_in *src, struct sockaddr_in *dest,
+			      struct usdf_dev_entry *dep)
+{
+	char dest_str[INET_ADDRSTRLEN];
+	char src_str[INET_ADDRSTRLEN];
+	char dev_str[INET_ADDRSTRLEN];
+	struct usd_device_attrs *dap;
+	int reachable;
+	int ret;
+
+	dap = &dep->ue_dattr;
+
+	/* Skip the device if it has problems. */
+	if (!dep->ue_dev_ok) {
+		USDF_WARN_SYS(FABRIC, "skipping %s/%s\n", dap->uda_devname,
+			     dap->uda_ifname);
+		return false;
+	}
+
+	/* If the given source address is not INADDR_ANY, compare against the
+	 * device.
+	 */
+	if (src && (src->sin_addr.s_addr != INADDR_ANY)) {
+		if (src->sin_addr.s_addr != dap->uda_ipaddr_be) {
+			inet_ntop(AF_INET, &src->sin_addr.s_addr, src_str,
+				  sizeof(src_str));
+			inet_ntop(AF_INET, &dap->uda_ipaddr_be, dev_str,
+				  sizeof(dev_str));
+			USDF_WARN_SYS(FABRIC, "src addr<%s> != dev addr<%s>\n",
+				      src_str, dev_str);
+			return false;
+		}
+	}
+
+	/* Check that the given destination address is reachable from the
+	 * interface.
+	 */
+	if (dest && dest->sin_addr.s_addr != INADDR_ANY) {
+		ret = usdf_get_distance(dap, dest->sin_addr.s_addr, &reachable);
+		if (ret) {
+			USDF_WARN_SYS(FABRIC, "usdf_get_distance failed\n");
```

Comment:
Can you expand this error message to include at least the destination?  If there's a problem with getting the distance, I think we need to know at least the destination to start to diagnose whatever the problem is.